onnxscript/ir/_core.py

```diff
@@ -389,89 +394,78 @@ def metadata_props(self) -> dict[str, str]:
 
-    def __index__(self) -> int:
-        return int(self)
+        Args:
+            value: The value of the dimension. It should not be an int.
+        """
+        if isinstance(value, int):
+            raise TypeError("The value of a SymbolicDim cannot be an int")
+        self._value = value
 
     def __eq__(self, other: object) -> bool:
-        if isinstance(other, (int, str)) or other is None:
+        if not isinstance(other, SymbolicDim):
             return self.value == other
-        if not isinstance(other, Dimension):
-            return False
         return self.value == other.value
 
-    def __ne__(self, value: object) -> bool:
-        return not self.__eq__(value)
-
-    def __lt__(self, other: object) -> bool:
-        if not isinstance(other, (int, Dimension)):
-            raise TypeError(f"Expected other to be Dimension or int, got {type(other)}")
-        return int(self) < int(other)
-
-    def __le__(self, other: object) -> bool:
-        if not isinstance(other, (int, Dimension)):
-            raise TypeError(f"Expected other to be Dimension or int, got {type(other)}")
-        return int(self) <= int(other)
-
-    def __gt__(self, other: object) -> bool:
-        if not isinstance(other, (int, Dimension)):
-            raise TypeError(f"Expected other to be Dimension or int, got {type(other)}")
-        return int(self) > int(other)
-
-    def __ge__(self, other: object) -> bool:
-        if not isinstance(other, (int, Dimension)):
-            raise TypeError(f"Expected other to be Dimension or int, got {type(other)}")
-        return int(self) >= int(other)
-
     def __hash__(self) -> int:
         return hash(self.value)
 
     @property
-    def value(self) -> int | str | None:
+    def value(self) -> str | None:
         return self._value
 
-    @property
-    def denotation(self) -> str | None:
-        return self._denotation
-
     def __str__(self) -> str:
         return f"{self._value}"
 
     def __repr__(self) -> str:
-        if self.denotation is not None:
-            denotation_text = f", denotation={self.denotation!r}"
-        else:
-            denotation_text = ""
-        return f"{self.__class__.__name__}({self._value}{denotation_text})"
+        return f"{self.__class__.__name__}({self._value})"
 
 
 class Shape(_protocols.ShapeProtocol, _display.PrettyPrintable):
-    __slots__ = ("_dims",)
-
-    def __init__(self, dims: _protocols.SimpleShape | Sequence[Dimension]) -> None:
-        # TODO: Support symbolic shapes with expressions?
-        for dim in dims:
-            if dim is not None and not isinstance(dim, (int, str, Dimension)):
-                raise TypeError(f"Expected int, str, None or Dimension, got '{type(dim)}'")
-        self._dims: list[Dimension] = [
-            dim if isinstance(dim, Dimension) else Dimension(dim) for dim in dims
+    __slots__ = ("_dims", "_frozen")
+
+    def __init__(
+        self,
+        dims: Iterable[int | SymbolicDim | Any],
```

Comment:
Why is `Any` here?

Reply:
Good catch - will remove

Reply:
Is it supposed to be `str | None` ?

Reply:
Actually it is because: When a ``dim`` is not an integer or a SymbolicDim, it is converted to a SymbolicDim.

Reply:
This allows users to create Shape(["a", "b"])

Reply:
Yes `str | None` is better. I was one step ahead and thinking about sympy objects. But we don't care about them now.

Reply:
Done